### PR TITLE
A few API changes based on comments and design work.

### PR DESCRIPTION
**Pull request by geeknoid**
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/15)

### CODE REVIEW

File: config/v1/metric_descriptor.proto [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+// A given metric is described by a set of labels. These labels represent the
+// different dimensions to associate with the metric. A given metric holds a
+// unique value for potentially any combination of these labels.
+//
+// Metric values can be integers, floating point numbers, strings, or booleans.
+//
+// Here's an example metric descriptor defined in a YAML configuration
+// file:
+//
+//    metrics:
```
Comment (mandarjog) [inline config/v1/metric_descriptor.proto]: I am thinking of a metric definition from the standpoint of entire istio, not just mixer.

```
- name: request_size
       elementary_measurement: request.size 
       metric_kind: DELTA
       value_type: INT64
       labels:
      - key: consumer_id
      - key: status_code
      - key: service_id
- name: bytes_transferred
       elementary_measurement: request.size + response.size
       metric_kind: DELTA
       value_type: INT64
       labels:
      - key: consumer_id
      - key: status_code
      - key: service_id
```
Disregarding proxy side aggregation, here is how this information is used jointly by mixer and the proxy.

proxy needs to know
1. The union of all labels to send across all metrics  [consumer_id, status_code, service_id]
2. The union of elementary_measurements needed by the metrics  [request.size, response.size]

Report call will contain all the above information. At this point mixer is able to assemble the actual metrics and add labels. 

.... However the same declarative metric definition can be used to fully compute the metric on proxy side.

Comment (mandarjog) [inline config/v1/metric_descriptor.proto]: + @qiwzhang this is what I was talking about this morning.

File: api/v1/report.proto [hunk 2]
```diff
@@ -31,11 +31,14 @@ message ReportRequest {
   // within the current stream.
   map<string, string> facts = 2;
 
+  // The monitored resource type associated with this report.
+  string monitored_resource_descriptor_name = 3
+
   // Metric values being reported
-  repeated MetricValue metric_values = 3;
+  repeated MetricValue metric_values = 4;
```
Comment (qiwzhang) [inline api/v1/report.proto]: Does this imply that the proxy will populate metrics and log_entries and send to mixer?  I think a better design is: the proxy just send the primary data, such as size, duration, ip, and mixer will populate the metrics and log-entries based on its backend plugins.

File: config/v1/metric_descriptor.proto (hunk 1 above)
Comment (qiwzhang) [inline config/v1/metric_descriptor.proto]: This looks good. 

But I prefer only mixer needs to know about this, no proxy.  The proxy just collects the facts (such as size, duration, ip) and send to mixer.

Comment (geeknoid) [inline config/v1/metric_descriptor.proto]: Mandar, what is "request.size" in your world? A fact, or a label produced by mapping facts, or something new?

In the API we have now, what I think you are trying to capture via elementary_measurement is handled imperatively by the caller. The MetricDescriptor type is really intended to describe the characteristic of a single metric value, not how these values are acquired. So the MetricDescriptor says "I want to hold integer values" and "I want to have different physical counters for each label I define". When someone calls Report, they provide the metric descriptor name, and a metric value (an integer). The mixer then derives the needed labels to properly 'route' the value, and forwards the result to the adapters.

I don't want to bind the definition of the metric with how the metric is acquired, those are separate concerns.

**Pull request merged**